Improve QSPIFBlockDevice conformance to SFDP

kyle-cypress opened the pull request:
### Description
Proposed fix for #11530 .

In several cases QSPIFBlockDevice does not conform to the SFDP standard [JESD216](https://www.jedec.org/system/files/docs/JESD216D-01.pdf).  These include:

 - Status register reading and writing
 - Soft resetting
 - Enabling of 4-byte addressing
 - Enabling of quad output mode
 - Use of the legacy erase instruction

In these cases QSPIFBlockDevice often assumes that certain instructions or methods are universal across flash devices when they are not.  This can cause issues in devices that do not match these assumptions.  This pull request resolves many of these issues and better aligns QSPIFBlockDevice with JESD216.
<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

This PR is dependent on #11604 

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

Commit:
Update reading/writing of status registers to conform to SFDP standard

Commit:
Reorder some functions

Commit:
Update flash device reset to conform to SFDP standard

Commit:
Enable 4-byte addressing when supported in accordance with the SFDP standard

Commit:
Clear block protection on non-SST flash devices

Commit:
Move configuration of QSPI format to within commands where it is necessary

Commit:
Update SDFP erase detection to properly handle legacy erase instruction

Commit:
Replace power function with bit shift

Commit:
Streamline setting of instruction member variables

Commit:
Add missing debug prints to command functions

Commit:
Correct typos and formatting

Commit:
Update QSPI format after enabling 4-byte addressing

Commit:
Enable some of the kvstore tests for PSoC 6 MCUs

Commit:
Remove redundant QSPI erase alignment

Commit:
Generalize KVStore phase 1/2 test BlockDevice sizes

Commit:
Enable TDBStore whitebox test on PSoC 6

Commit:
Update QSPI test to reflect fixes in QSPIFBlockDevice

Commit:
Fix Astyle issues

0xc0170 commented on components/storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.cpp:
```diff
@@ -380,7 +380,7 @@ int QSPIFBlockDevice::erase(bd_addr_t addr, bd_size_t in_size)
     uint32_t offset = 0;
     uint32_t chunk = 4096;
-    unsigned int cur_erase_inst = _erase_instruction;
+    qspi_inst_t cur_erase_inst = _erase_instruction;
```
What is the reason for this change? I dont see  anything in the commit msg to explain.

kyle-cypress replied:
Consistency - the current code uses a mixture of signed and unsigned integers to represent instruction values
Clarity - convey the meaning better of the data than a generic `int` (and encourage consistency)
Efficiency: The instruction is only 8 bits wide, so size the data type to fit.

Commit:
Remove hard-coded instruction ids from QSPI Tests

Replace with macros from the test flash_config header, consistent with
how most commands are built in this test.

0xc0170 commented on drivers/QSPI.h:
```diff
@@ -160,7 +166,7 @@ class QSPI : private NonCopyable<QSPI> {
      *    Returns QSPI_STATUS_SUCCESS on successful reads and QSPI_STATUS_ERROR on failed reads.
      */
-    qspi_status_t read(int instruction, int alt, int address, char *rx_buffer, size_t *rx_length);
+    qspi_status_t read(qspi_inst_t instruction, int alt, int address, char *rx_buffer, size_t *rx_length);
```
`int` should be sufficient for inst as it can be any number, narrow down this won't make much difference here.

Same as alt, address and others.

This is breaking change, so needs to be justified to go in.

0xc0170 replied:
The usage of this in the block device was wrong, should follow what the driver specifies.

kyle-cypress replied:
> The usage of this in the block device was wrong, should follow what the driver specifies.

Can you clarify what you mean? Are you saying that the use of `unsigned int` instead of `int` in the block device was wrong?

kyle-cypress replied:
> `int` should be sufficient for inst as it can be any number, narrow down this won't make much difference here.
> 
> Same as alt, address and others.
> 
> This is breaking change, so needs to be justified to go in.

This change is not mandatory for the fix here so I can pull it out of this review and into a separate PR if it makes things easier. But in principle, shouldn't values which correspond to bit patterns of a particular width be represented as types of particular widths? If so, then I would argue that a separate PR should also replace  `alt` and `address` with more specific types.

SeppoTakalo replied:
I'm not that familiar with SQPIF, but if commands are limited to 8 bytes, then I would say that this is acceptable, and it is not a breaking change.

Mbed OS have not guaranteed to have binary compatibility, so moving from int to uint8_t should be OK.

SeppoTakalo reviewed (changes requested):
There are lots of changes not related to QSPIF, can you split your changes to separate PRs so that we can properly review each change on its own.

Currently I'm not able to understand why this PR touches things like LittleFS, KVStore and DeviceKey.

SeppoTakalo commented on components/storage/blockdevice/COMPONENT_QSPIF/QSPIFBlockDevice.cpp:
```diff
@@ -132,23 +181,24 @@ int QSPIFBlockDevice::init()
     _mutex.lock();
 
+    // All commands other than Read and RSFDP use default 1-1-1 bus mode (Program/Erase are constrained by flash memory performance more than bus performance)
+    _qspi.configure_format(QSPI_CFG_BUS_SINGLE, QSPI_CFG_BUS_SINGLE, _address_size, QSPI_CFG_BUS_SINGLE,
```
Do you need to check return value from `configure_format()`?

kyle-cypress replied:
Yes, the return value needs to be checked for configure_format across the board, but that is handled in #11603 (applying the check to additional/relocated configure_format calls would be handled during the rebase of whichever PR goes in second).

Commit:
QSPIF: Add back enable_fast_mode

This function writes a "config" register to ensure that the flash part
is in high performance mode, not low-power mode. This is required at
by at least MX25R6435F in order to operate at frequencies > 33MHz
(for reference, DISCO_L475VG_IOT01A runs the QSPI interface at 80 MHz).
The config register that this writes does not appear to be covered by
the SFDP spec (JESD216D.01) so this remains the status quo of
unconditional execution, as has been done on master since #8352.

Commit:
Don't clear quad enable when clearing block protection

QSPIFBlockDevice::_clear_block_protection() has logic to retain the
WIP and WEL bits in status register 1, but it failed to account for
the situation where the QE bit is also in status register 1.
In _sfdp_set_quad_enabled, note the status register and bit therein
for the quad enable, so that _clear_block_protection can retain it.

Commit:
QSPIF: Centralize handling of vendor quirks

Introduce a separate function for handling alterations to device interaction
which are not covered by the SFDP tables and therefore require checking against
the vendor id.

Commit:
QSPIF: Handle parts with extra config registers

Default to 2 status registers, but update this value if necessary
 during vendor quirk handling for parts (currently only Macronix)
 which have one status register and two control registers. For the
 purposes of QSPIFBlockDevice, these are all considered status
 (or at least "status-like") registers because they are all written
 via the Write Status Register instruction.
Set the custom RDCR instruction for Macronix during quirk handling.
Update reading and writing of status registers to handle a variable
 number of status registers.

Commit:
QSPIF: Handle fast mode enable via vendor quirks

Use a vendor id check to only perform this enable on devices which define the
 second configuration register where the fast mode enable bit lives.
Change _enable_fast_mode to use the standard status register reading and writing functions

Commit:
Disable attempted 4-byte addressing for some boards

4-byte addressing has been seen to cause failures on NORDIC
boards and with Macronix memories. Suppress the attempt to enable it
on that hardware (via vendor quirks and a target check) until either
the failure cause can be fixed or a more robust suppression mechanism
is implemented.

Commit:
General Block Device Test: Expand Thread Stack

The addition of trace logging during greentea tests pushes the multithreaded
read-write test beyond the limits of the stack it allocates for its threads.
The increase of 128 bytes was chosen by experimentation.
